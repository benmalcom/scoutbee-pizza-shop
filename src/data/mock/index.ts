import { v4 as uuid } from 'uuid';
import { sample } from 'lodash';

export interface PizzaIngredient {
  name: string;
  id: string;
}

export interface Pizza {
  id?: string;
  name: string;
  status: string;
  ingredients: PizzaIngredient[];
  timestamp: number
}

const pizzaNames: string[] = [
  'Margherita', 'Marinara', 'Quattro Stagioni', 'Carbonara', 'Frutti di Mare',
  'Romana', 'Fattoria', 'Schiacciata', 'Prosciutto', 'Quattro Formaggi', 'Crudo',
  'Napoletana', 'Pugliese', 'Montanara', 'Emiliana', 'Americana', 'Sarda', 'Tonno',
];

export const pizzaStatuses: string[] = ['Waiting', 'Ready'];

export const pizzaIngredients: PizzaIngredient[] = [
  { name: 'Cheese', id: uuid() },
  { name: 'Bacon', id: uuid() },
  { name: 'Mushrooms', id: uuid() },
  { name: 'Ananas', id: uuid() },
  { name: 'Tomato sauce', id: uuid() },
  { name: 'Mozzarella', id: uuid() },
  { name: 'Parmesan', id: uuid() },
  { name: 'Eggs', id: uuid() },
  { name: 'Bacon', id: uuid() },
];

export const createMockPizzaList = () => {
  // Generate 30 mock pizza
  const items: Pizza[] = [];
  for (let i = 0; i < 30; i++) {
    const name: string = sample(pizzaNames) || pizzaNames[0];
    const status: string = sample(pizzaStatuses) || pizzaStatuses[0];
    const randomSize =  Math.floor(Math.random() * (pizzaIngredients.length - 1 + 1)) + 1;
    items.push({
      id: uuid(),
      name,
      status,
      ingredients: pizzaIngredients.slice(0, randomSize),
      timestamp: Date.now()
    });
  }
  return items;
};

