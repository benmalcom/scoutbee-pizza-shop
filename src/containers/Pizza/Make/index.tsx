import React, { useContext } from 'react';
import Select from 'react-select';
import { useForm, Controller } from 'react-hook-form';
import { useHistory } from 'react-router-dom';
import { v4 as uuid } from 'uuid';
import './Make.scss';
import { Pizza, PizzaIngredient, pizzaIngredients } from '../../../data/mock';
import { PizzaItemContext } from '../../../components/PizzaContext';

const Make = () => {
  const { register, errors, handleSubmit, control } = useForm();
  const { setItems } = useContext(PizzaItemContext);
  const history = useHistory();

  const onSubmit = (data: any) => {
    const { name, ingredients = [] } = data;
    const payload: Pizza = {
      status: 'Waiting',
      id: uuid(),
      timestamp: Date.now(),
      name,
      ingredients: ingredients.map(({ label, value }: any)  => ({ name: label, id: value }))
    };
    setItems((current: Pizza[]) => [payload, ...current]);
    history.push('/');
  };

  const options = pizzaIngredients.map((item: PizzaIngredient) => ({ label: item.name, value: item.id }));

  return <div className="Make shadow-sm p-5 bg-white rounded col-xl-5 col-lg-5 col-md-6 col-sm-8 col-11 mx-auto mt-5">
    <h3 className="mb-4">Make Pizza</h3>

    <form data-testid="form" onSubmit={handleSubmit(onSubmit)}>
      <div className="form-group">
        <label>Pizza name</label>
        <input data-testid="name" name="name" className="form-control" placeholder="Enter name..." ref={register({
          required: 'This name is required.',
          minLength: {
            value: 5,
            message: 'This input must exceed 5 characters',
          },
        })}/>
        {errors.firstName && <small className="form-text text-danger">{errors.firstName.message}</small>}
      </div>
      <div className="form-group">
        <label htmlFor="ingredients">Ingredients</label>
        <Controller
          control={control}
          render={({ onChange, onBlur, value }): any => (
            // @ts-ignore
            <Select
              inputId="ingredients"
              blurInputOnSelect={false}
              isClearable
              hideSelectedOptions
              placeholder="Select..."
              classNamePrefix="react-select"
              isSearchable
              isMulti
              options={options}
              value={value}
              onChange={(selected: PizzaIngredient[]) => onChange(selected)}
              onBlur={onBlur}
            />

          )}
          name="ingredients"
          rules={{ required: 'Add at least one ingredient' }}
        />
        {errors.ingredients && <small className="form-text text-danger">{errors.ingredients.message}</small>}
      </div>
      <button id="submit" type="submit" className="btn btn-info">Submit</button>
    </form>

  </div>;
};

export default Make;
