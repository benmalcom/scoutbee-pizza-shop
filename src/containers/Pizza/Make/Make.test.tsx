import React from 'react';
import { render, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import selectEvent from 'react-select-event';
import { shallow } from 'enzyme';
import { pizzaIngredients } from '../../../data/mock';
import Make from './';


const mockData = {
  pizzaName: 'Pepperoni',
  pizzaIngredients: pizzaIngredients[0],
};
afterEach(cleanup);

test('<Make/>  renders successfully', () => {
  const component = shallow(<Make/>);
  expect(component).toMatchSnapshot();
});

test('<Make/> renders page title', () => {
  const { getByText } = render(<Make/>);
  const homeTitle = getByText(/Make Pizza/i);
  expect(homeTitle).toBeInTheDocument();
});

test('<Make/> should watch input correctly', () => {
  const { getByTestId, getByText, getByLabelText } = render(<Make/>);

  const nameInput = getByTestId('name');

  userEvent.type(nameInput, mockData.pizzaName);
  expect(nameInput).toHaveValue(mockData.pizzaName);

  selectEvent.openMenu(getByLabelText('Ingredients'));
  expect(getByText(mockData.pizzaIngredients.name)).toBeInTheDocument();

});

