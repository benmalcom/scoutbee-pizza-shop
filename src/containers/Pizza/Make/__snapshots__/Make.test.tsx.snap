// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Make/>  renders successfully 1`] = `ShallowWrapper {}`;
